### ISSUE TITLE
Configure tab size when viewing a file

### ISSUE THREAD

**Issue by markusamshove** [go-gitea/gitea#9071]
Hi,

Is there an option to configure how many spaces a tab contains?

Our java files are indented with tabs and when I view them, every tab seems to be 8 spaces wide.
Normally I would expect 4 spaces for java, c# and similar languages

**Comment by silverwind** [comment]
I think you can put a [`.editorconfig`](https://editorconfig.org/) in your repo and some aspects like the editor will honour it, not sure about repo viewer.

**Comment by KaKi87** [comment]
Hello,

How can I change Gitea's global tab size from 8 to 4 ?

Thanks.

**Comment by Spongman** [comment]
please reopen.

**Comment by KaKi87** [comment]
up

**Comment by KaKi87** [comment]
u

**Comment by Exagone313** [comment]
I build Gitea with this patch:
```diff
diff --git a/modules/templates/helper.go b/modules/templates/helper.go
index e4107dfa9..38c4c661f 100644
--- a/modules/templates/helper.go
+++ b/modules/templates/helper.go
@@ -185,18 +185,18 @@ func NewFuncMap() []template.FuncMap {
 			)
 			if ec != nil {
 				if value, ok = ec.(*editorconfig.Editorconfig); !ok || value == nil {
-					return "tab-size-8"
+					return "tab-size-4"
 				}
 				def, err := value.GetDefinitionForFilename(filename)
 				if err != nil {
 					log.Error("tab size class: getting definition for filename: %v", err)
-					return "tab-size-8"
+					return "tab-size-4"
 				}
 				if def.TabWidth > 0 {
 					return fmt.Sprintf("tab-size-%d", def.TabWidth)
 				}
 			}
-			return "tab-size-8"
+			return "tab-size-4"
 		},
 		"SubJumpablePath": func(str string) []string {
 			var path []string
```

**Comment by tmont** [comment]
I had to add this to @Spongman's custom template to get things working in Firefox as it [does not (yet) support](https://caniuse.com/?search=tab-size) un-prefixed `tab-size`:

```html
<style>
.tab-size-8 {
        tab-size: 4 !important;
        -moz-tab-size: 4 !important;
}
</style>

**Comment by vhscom** [comment]
Here's a bookmarklet you can use to format the tab size for Gitea when viewing code in compare/diff view through the web interface.

```js
javascript:document.querySelectorAll('table.chroma').forEach((el)=>el.style.tabSize=2)
```

Adjust to your liking. Alternative options are to (a) add an `.editorconfig` file to the repository [as mentioned](https://github.com/go-gitea/gitea/issues/14225#issuecomment-753628646) by @silverwind or (b) use a different diff viewer [such as delta](https://github.com/dandavison/delta).

This issue should probably be closed as it has already been answered.

**Comment by KaKi87** [comment]
Well, you'd better use a userstyle rather than a bookmarklet

**Comment by silverwind** [comment]
I think this issue is still open because the code view doesn't respect the configured tab-size in `.editorconfig`. The web editor itself certainly already does.

**Comment by KaKi87** [comment]
Requiring such file doesn't seem right.

**Comment by silverwind** [comment]
Yes, I think a user setting for default tab size is also desirable. But `.editorconfig` should have precedence when present.

**Comment by MarkusAmshove** [comment]
I'd say:

User settings > editorconfig > highlighting lib > default

**Comment by danebou** [comment]
Who the heck uses a tab size of 8 by default? lol. Really surprised this isn't fixed.

**Comment by vhscom** [comment]
Do people still use Editor Config now that Prettier is a thing? Honestly I opened up the docs site for the first time in, um, maybe 5 years (no 6 years) and tried to set one up just so I could point people at my code in Codeberg and I decided I'd rather eat paint chips than put another dotfile in my project root.

**Comment by vhscom** [comment]
One thing to take into consideration as that I doubt many projects do, or would even want, to add an `.editorconfig` file if they're using prettier. I could see that causing headaches if they're not in perfect alignment. Besides, who wants to add another dotfile to their repo. A sensible default of 4 spaces would be fine. But it would be even better to give the user control straight from the interface.

**Comment by vhscom** [comment]
One thing to take into consideration is that I doubt many projects do, or would even want, to add an `.editorconfig` file if they're using prettier. I could see that causing headaches if they're not in perfect alignment. Besides, who wants to add another dotfile to their repo. A sensible default of 4 spaces would be fine. But it would be even better to give the user control straight from the interface.

**Comment by KaKi87** [comment]
Yes. Personally, I started using spaces as indentation character instead of tabs because of this issue. And this isn't right.

**Comment by silverwind** [comment]
> Who the heck uses a tab size of 8 by default? lol. Really surprised this isn't fixed. In addition to this, the default tab size should be 4 since almost every IDE out of the box uses 4.

FWIW, Browsers are dictated by spec to use [default size 8](https://drafts.csswg.org/css-text/#tab-size-property) thought I do agree 4 is a generally better default.

**Comment by delvh** [comment]
While I can definitely see why adding it as a query param would be useful, how would it handle the whitespace for files with differing behavior defined in the `editorconfig`?
Should the editorconfig simply be ignored globally when that parameter is set?

**Comment by silverwind** [comment]
I think in general we want a setting UI where user can set their preference, similar to GitHub now has for tab size:

https://github.com/settings/appearance

That setting should overrule `.editorconfig`. Only through the special value `default` would editorconfig then be applied.

**Comment by Crystal-RainSlide** [comment]
> While I can definitely see why adding it as a query param would be useful, how would it handle the whitespace for files with differing behavior defined in the `editorconfig`? Should the editorconfig simply be ignored globally when that parameter is set?

> That setting should overrule `.editorconfig`. Only through the special value `default` would editorconfig then be applied.

On GitHub: `ts` parameter > `.editorconfig` > tab size in the user settings. Everyone can check this by themselves.

This seems to be decided by the scope. `ts` parameter is specifc to a webpage, `.editorconfig` is specifc to a repo, and tab size in the user settings is specifc to everything a user sees, and the smaller the scope the higher the priority.

---

> I think in general we want a setting UI where user can set their preference, similar to what GitHub now has for tab size:
>
> [github.com/settings/appearance](https://github.com/settings/appearance)

Yes.

---

> We can du the url param too, but I think it's too obscure of a feature in general to be useful.

Obscure is what GitHub had done wrong, anyone else can get it right (GitHub itself included).

I'd already talked about the "Change tab size" button thing, but let's see if I can come up with some showcase...

**Comment by Spongman** [comment]
it should be simple for the user to easily, and permanently override whatever is in a repo's .editorconfig. the whole point of using tabs is that the indentation amount is configurable by the viewer, _not_ baked into the source.

**Comment by omenos** [comment]
Both GitHub and GitLab have a user preference for defining the tabstop width for _viewing_ files. As long as the `.editorconfig` only applies to the editor view, it shouldn't be viewed as a solution here and other related issues.

Tabstop size is important for languages that enforce tabs over spaces, such as Go. I'm currently using the Stylus extension to override tabstops for forge sites, but this is a browser specific workaround on my desktop; it does not apply to mobile.

**Comment by silverwind** [comment]
I guess we can implement 2,4,8,auto option, default being `auto` which derives from editorconfig, and possibly falls back to a default of 4 (or 8, which is still what browsers default to, but I think it's antiquated and 4 is a better default).

**Comment by arkadiuszmakarenko** [comment]
> I guess we can implement `2`,`4`,`8`,`auto` options. Default being `auto` which derives from editorconfig, and possibly falls back to a default of 4 (or 8, which is still what browsers default to, but I think it's antiquated and 4 is a better default).

This would be very welcomed change.

**Comment by eliliam** [comment]
How's this coming? Six months ago @silverwind indicated the setting could be added, has anyone been able to work on this yet?

**Comment by Palid** [comment]
@silverwind do you need help with someone taking this over? I'd gladly implement this change

**Comment by silverwind** [comment]
Feel free to take a step. Personally I don't need this feature because I'm a strong proponent of `.editorconfig`.

**Comment by jclsn** [comment]
Would like to see this too. Such basic feature!

**Comment by KaKi87** [comment]
[Forgejo](https://forgejo.org/) supports this btw

**Comment by omenos** [comment]
> [Forgejo](https://forgejo.org/) supports this btw

Which part? Forgejo does default to a width of four characters for tabs, but as of Forgejo v10 I have not seen a user configurable tab size parameter available. I may have missed it, but there is a similar issue over on Forgejo about this same topic:

[feat: tab size should be configurable #5672](https://codeberg.org/forgejo/forgejo/issues/5672)

**Comment by KaKi87** [comment]
Sorry for the ambiguity, I meant it supports `.editorconfig`, which is the majority request here.

**Comment by wxiaoguang** [comment]
> Sorry for the ambiguity, I meant it supports `.editorconfig`, which is the majority request here.

But forjoge ever largely copied Gitea's code, and new Gitea has many unique features that are not in forgejo.

For `editorconfig` part, there is no difference.

**Comment by KaKi87** [comment]
Mea culpa, since users here requested for the file viewer to support `.editorconfig` but no one ever announced here this was implemented, I thought when seeing in that linked Forgejo issue someone saying it was implemented, that Forgejo maintainers did it. Sorry for the disturbance.

So it's still good to know for the people here though, that now they can configure tab size per-repo with `.editorconfig`.

Have a ice week

**Comment by wxiaoguang** [comment]
> Mea culpa, since users here requested for the file viewer to support `.editorconfig` but no one ever announced here this was implemented, I thought when seeing in that linked Forgejo issue someone saying it was implemented, that Forgejo maintainers did it. Sorry for the disturbance.

That code was also taken from Gitea.

For the author's question: 

* "Normally I would expect 4 spaces for java, c# and similar languages": it has been resolved, because now 4 is the default width.
* To fully customize the tab width: it needs to involve the user setting related modules, and maybe we should refactor the user setting package first.

**Comment by omenos** [comment]
> (DONE) To use repository-related config: `editorconfig` is supported

As the idea was mentioned up in the Forgejo discussion, does Gitea supported rendering projects using [nested `editorconfig`](https://learn.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options#file-hierarchy-and-precedence) setups? I can see this as a useful feature for multi-language monorepos.

**Comment by silverwind** [comment]
I don't think a user setting is strictly needed as preferred tab size can differ based on project. One project might render better with 4, another one with 8. I think we could simply add `<select>` elements to the file view and code editor that changes the tab size only on the current page, not remembering anything.

**Comment by wxiaoguang** [comment]
> does Gitea supported rendering projects using [nested `editorconfig`](https://learn.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options#file-hierarchy-and-precedence) setups? I can see this as a useful feature for multi-language monorepos.

I think no at the moment.

It's not difficult to support this feature, the only concern is that it might slow down some pages slightly because it needs to check every parent dir to detect `editorconfig` and apply.

So IMO the issue looks like this at the moment:

1. The current 4-width tab should be good enough for most cases.
1. If there is a strong requirement to support `editorconfig` in mono-repo (many users need it):
    * If it doesn't slow down the page too much, just do it.
    * Otherwise, there could be a config option to enable or disable it.

**Comment by silverwind** [comment]
> It's not difficult to support this feature, the only concern is that it might slow down some pages slightly because it needs to check every parent dir to detect editorconfig and apply.

I'm not sure how the editorconfig module does it, but ideally this is a O(n) operation, n being the number of parent directories to the viewed file.

**Comment by silverwind** [comment]
I will look into adding these `<select>` boxes soon to file view and code editor.

**Comment by adamel** [comment]
Personally I would very much have the ability to configure a global .editorconfig config in Gitea to use for all repositories that does not have one.

We have different indent settings for different languages, but all repositories use the same styles, so adding a .editorconfig file to each of several hundred repositories does not really make sense.

**Comment by silverwind** [comment]
Makes sense to have such a file, but it may become challenging to merge multiple `.editorconfig` files, maybe https://github.com/editorconfig/editorconfig-core-go has some helpers for that thought.

**Comment by adamel** [comment]
For us there is no need to merge multiple files - if there is a file in the repository use only that, otherwise use only the globally configured file from Gitea as a fallback.

**Comment by silverwind** [comment]
Right, could just take the global file as a "fallback".
